File: Algorithms/C++/00448-Find All Numbers Disappeared in an Array.cpp
class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        
        int n = nums.size();

        // negate num at index to mark seen
        for(int i{0}; i < n; i++){
            int index = std::abs(nums[i]) - 1;
            nums[index] = std::abs(nums[index]) * -1;
        }

        std::vector<int> ans;
        for(int i{0}; i < n; i++){
            if(nums[i] > 0){
                ans.push_back(i+1);
            }
        }

        return ans;
    }
};

/*
    O(n)
    O(1)

    use the index
     0 1 2 3
    [1,4,4,4]
*/

File: Algorithms/Python3/00448-Find All Numbers Disappeared in an Array.py
class Solution:
    def findDisappearedNumbers(self, nums: List[int]) -> List[int]:
        hashset = set(nums)
        result = []
        for i in range(1, len(nums)+1):
            if i not in hashset:
                result.append(i)
        
        return result

'''
    tc: O(n)
    sc: O(n)
    put numbers into an unordered set
    run through 1 to n to check for missing numbers in set

    tc: O(n)
    sc: O(1)
    use indices to check if number is missing
    use swaps because of inplace solution

    iterate through nums and swap number with its index
    
    [4,3,2,7,8,2,3,1]
    [7,3,2,4,8,2,3,1]

    [1,1]
    [-1,1]
'''